### PR TITLE
Implement/Update POSIX mutex APIs

### PR DESCRIPTION
**Pull request by punitvara**
Add

### CODE REVIEW

**Comment by punitvara** [comment]
I will add more code to test case and mutex.c. Pushed early so that  I can fix review comment along with developing rest of code.

File: lib/posix/pthread_mutex.c [hunk 1]
```diff
@@ -9,6 +9,18 @@
...
 
+/* Memory pool for pthread space */
+K_MEM_POOL_DEFINE(posix_mutex_sem, sizeof(struct k_sem),
+		  sizeof(struct k_sem), 32, 4);
+
+/*
+ *  Default mutex attrs.
+ */
+static pthread_mutexattr_t def_attr = {
+	PTHREAD_MUTEX_DEFAULT, /* type */
```
Comment (lpereira) [inline lib/posix/pthread_mutex.c]: Please define this as `static const` and use named initializers: `{ .type = PTHREAD_MUTEX_DEFAULT, .protocol = PTHREAD_PRIO_NONE }`.

File: lib/posix/pthread_mutex.c [hunk 2]
```diff
@@ -9,6 +9,18 @@
 #include <wait_q.h>
 #include <posix/pthread.h>
 
+/* Memory pool for pthread space */
+K_MEM_POOL_DEFINE(posix_mutex_sem, sizeof(struct k_sem),
+		  sizeof(struct k_sem), 32, 4);
```
Comment (lpereira) [inline lib/posix/pthread_mutex.c]: Shouldn't this use a `k_mutex` instead of a `k_sem`? There are some subtle differences that will make it harder to implement by using `k_sem`.  For instance, depending on the mutex type, threads may lock mutexes if they have been already locked.  Please see the man page for `pthread_mutex_lock()` for a table with the behavior for each type:

```
                   ┌───────────┬────────────┬────────────────┬───────────────────────┐
                   │Mutex Type │ Robustness │     Relock     │ Unlock When Not Owner │
                   ├───────────┼────────────┼────────────────┼───────────────────────┤
                   │NORMAL     │ non-robust │ deadlock       │ undefined behavior    │
                   ├───────────┼────────────┼────────────────┼───────────────────────┤
                   │NORMAL     │ robust     │ deadlock       │ error returned        │
                   ├───────────┼────────────┼────────────────┼───────────────────────┤
                   │ERRORCHECK │ either     │ error returned │ error returned        │
                   ├───────────┼────────────┼────────────────┼───────────────────────┤
                   │RECURSIVE  │ either     │ recursive      │ error returned        │
                   │           │            │ (see below)    │                       │
                   ├───────────┼────────────┼────────────────┼───────────────────────┤
                   │DEFAULT    │ non-robust │ undefined      │ undefined behavior†   │
                   │           │            │ behavior†      │                       │
                   ├───────────┼────────────┼────────────────┼───────────────────────┤
                   │DEFAULT    │ robust     │ undefined      │ error returned        │
                   │           │            │ behavior†      │                       │
                   └───────────┴────────────┴────────────────┴───────────────────────┘

       †     If the mutex type is PTHREAD_MUTEX_DEFAULT, the behavior of pthread_mutex_lock() may
             correspond to one of the three other standard mutex types as described in the  table
             above.  If  it  does not correspond to one of those three, the behavior is undefined
             for the cases marked †.

```

File: lib/posix/pthread_mutex.c [hunk 3]
```diff
@@ -22,3 +34,97 @@ int pthread_mutex_trylock(pthread_mutex_t *m)
...
+}
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_destroy(pthread_mutex_t *m)
+{
+	ARG_UNUSED(m);
```
Comment (lpereira) [inline lib/posix/pthread_mutex.c]: Shouldn't the memory allocated by pthread_mutex_init() be returned to the pool?

File: lib/posix/pthread_mutex.c [hunk 4]
```diff
@@ -22,3 +34,97 @@ int pthread_mutex_trylock(pthread_mutex_t *m)
 
 	return ret;
 }
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_init(pthread_mutex_t *m,
```
Comment (lpereira) [inline lib/posix/pthread_mutex.c]: Will we support PTHREAD_MUTEX_INITIALIZER?

**Review by lpereira — CHANGES_REQUESTED** [review]

File: lib/posix/pthread_mutex.c (hunk 3 above)
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: Done

File: lib/posix/pthread_mutex.c (hunk 2 above)
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: No. I have forwarded you a mail which we discussed with Andy Ross for this.

File: tests/posix/mutex/src/posix_mutex.c [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	pthread_attr_setschedparam(&attr, &schedparam);
+
+	pthread_mutexattr_settype(&mut_attr, PTHREAD_MUTEX_NORMAL);
+	pthread_mutexattr_setprotocol(&mut_attr, PTHREAD_PRIO_NONE);
+	pthread_mutex_init(&mutex1, &mut_attr);
+
+	pthread_mutexattr_gettype(&mut_attr, &type);
+	pthread_mutexattr_getprotocol(&mut_attr, &protocol);
+
+	TC_PRINT("Type = %d Protocol = %d ", type, protocol);
```
Comment (ramakrishnapallala) [inline tests/posix/mutex/src/posix_mutex.c]: instead of TC_PRINT you can use zassert_true here

File: tests/posix/mutex/src/posix_mutex.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+#define STACK_SIZE 1024
+
+static K_THREAD_STACK_DEFINE(stack, STACK_SIZE);
+
+pthread_mutex_t mutex1;
+
+void *foo(void *p1)
+{
+	pthread_mutex_lock(&mutex1);
```
Comment (ramakrishnapallala) [inline tests/posix/mutex/src/posix_mutex.c]: you are simply using the API's without testing anything. you can create synchronization scenario b/w main thread and child thread.

File: lib/posix/pthread_mutex.c [hunk 7]
```diff
@@ -9,16 +9,183 @@
...
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_unlock(pthread_mutex_t *m)
+{
+	if (m->owner != pthread_self()) {
+		return EPERM;
+	}
+
+	m->lock_count--;
```
Comment (ramakrishnapallala) [inline lib/posix/pthread_mutex.c]: k_sem_give(m->sem);
m->lock_count--;
if (m->lock_count == 0) {
m->owner = NULL;
}

File: lib/posix/pthread_mutex.c [hunk 8]
```diff
@@ -9,16 +9,183 @@
...
+{
+	*type = attr->type;
+	return 0;
+}
+
+int pthread_mutexattr_settype(pthread_mutexattr_t *attr, int type)
+{
+	int retc = EINVAL;
+
+	if ((type == PTHREAD_MUTEX_NORMAL) ||
```
Comment (ramakrishnapallala) [inline lib/posix/pthread_mutex.c]: follow the same for setprotocol functiona also.

**Review by ramakrishnapallala — APPROVED** [review]
LGTM

**Comment by punitvara** [comment]
@laperie Can you please review this ? If looks then we can merge this one.

**Comment by punitvara** [comment]
@lpereira Can you please review this ? If looks then we can merge this one.

File: lib/posix/pthread_mutex.c [hunk 9]
```diff
@@ -9,16 +9,214 @@
...
+	if (m->lock_count == 0) {
+		m->owner = pthread_self();
+	}
+
+	if (m->owner == pthread_self()) {
+		m->lock_count++;
+		if (m->lock_count > MUTEX_MAX_REC_LOCK) {
+			return EAGAIN;
+		}
+		return k_sem_take(m->sem, timeout);
```
Comment (lpereira) [inline lib/posix/pthread_mutex.c]: k_sem_take() returns a negative error number or 0 on success. You have to convert this to a positive error code or 0: `return -k_sem_take(m->sem, timeout)`

File: lib/posix/pthread_mutex.c [hunk 10]
```diff
@@ -9,16 +9,214 @@
...
+		if (m->type == PTHREAD_MUTEX_ERRORCHECK) {
+			return EDEADLK;
+		}
+	}
+
+	if (m->type == PTHREAD_MUTEX_RECURSIVE) {
+		rc = mutex_recursive_take(m, timeout);
+	} else {
+		m->lock_count = 1;
+		rc = k_sem_take(m->sem, timeout);
```
Comment (lpereira) [inline lib/posix/pthread_mutex.c]: acquire_mutex() is called while the IRQs are locked.  If one passes a timeout of K_FOREVER, and the semaphore is taken, the system (and not the thread!) will freeze forever.

**Review by lpereira — CHANGES_REQUESTED** [review]

File: lib/posix/pthread_mutex.c [hunk 11]
```diff
@@ -9,16 +9,214 @@
...
+		return k_sem_take(m->sem, timeout);
+	}
+
+	return EPERM;
+}
+
+static int acquire_mutex(pthread_mutex_t *m, int timeout)
 {
-	int key = irq_lock(), ret = EBUSY;
+	pthread_t this_thread;
```
Comment (lpereira) [inline lib/posix/pthread_mutex.c]: Please do not mix variable declarations and initializations; just initialize `this_thread` as it's declared

File: lib/posix/pthread_mutex.c (hunk 9 above)
Comment (lpereira) [reply lib/posix/pthread_mutex.c]: Also, same comment about this function being called with IRQ locked and timeout being possibly K_FOREVER

File: lib/posix/pthread_mutex.c [hunk 12]
```diff
@@ -9,16 +9,214 @@
...
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_init(pthread_mutex_t *m,
+				     const pthread_mutexattr_t *attr)
+{
+	const pthread_mutexattr_t *mattr;
+
+	if (k_mem_pool_alloc(&posix_mutex_sem, &m->block,
+			sizeof(struct k_sem), 100) == 0) {
```
Comment (lpereira) [inline lib/posix/pthread_mutex.c]: Why is timeout `100`?

File: lib/posix/pthread_mutex.c [hunk 13]
```diff
@@ -9,16 +9,214 @@
...
+	if (m->owner != pthread_self()) {
+		return EPERM;
+	}
+
+	if (m->lock_count == 0) {
+		return -1;
+	}
+
+	k_sem_give(m->sem);
+	m->lock_count--;
```
Comment (lpereira) [inline lib/posix/pthread_mutex.c]: This is racy. The decrement and comparison should happen while the semaphore is taken.

File: lib/posix/pthread_mutex.c [hunk 14]
```diff
@@ -9,16 +9,214 @@
...
+ */
+int pthread_mutexattr_setprotocol(pthread_mutexattr_t *attr, int protocol)
+{
+	if ((protocol != PTHREAD_PRIO_NONE) &&
+	    (protocol != PTHREAD_PRIO_INHERIT) &&
+	    (protocol != PTHREAD_PRIO_PROTECT)) {
+		return EINVAL;
+	}
+
+	attr->protocol = protocol;
```
Comment (lpereira) [inline lib/posix/pthread_mutex.c]: The protocol can be set, but nothing checks it.  Please only support one protocol type that matches the k_sem behavior, and return EINVAL for all other types.

File: lib/posix/pthread_mutex.c [hunk 15]
```diff
@@ -9,16 +9,214 @@
...
+ * See IEEE 1003.1
+ */
+int pthread_mutex_unlock(pthread_mutex_t *m)
+{
+	if (m->owner != pthread_self()) {
+		return EPERM;
+	}
+
+	if (m->lock_count == 0) {
+		return -1;
```
Comment (lpereira) [inline lib/posix/pthread_mutex.c]: Is -1 a valid return value for pthread_mutex_unlock()?

**Comment by punitvara** [comment]
@lpereira I have updated changes. Can you please have a look ?

File: lib/posix/pthread_mutex.c (hunk 11 above)
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: Done

File: lib/posix/pthread_mutex.c (hunk 12 above)
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: There is not standard timeout. I just took inspiration from pthread.c where same constant has been used. However, I have created Macro to change this value in future if needed.

File: lib/posix/pthread_mutex.c (hunk 13 above)
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: http://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_mutex_lock.html 

I maintained separate tracing for lock count. According to definition, pthread_mutex_lock increases count and while unlocking mutex it decrease count. That why I have used it separately.

File: lib/posix/pthread_mutex.c (hunk 14 above)
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: Right only one protocol is supported. But in future other protocols can be added. So I added comment accordingly

File: lib/posix/pthread_mutex.c (hunk 13 above)
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: Still let me know if I am missing any of comments to be resolved

File: include/posix/pthread.h [hunk 16]
```diff
@@ -157,42 +159,61 @@ static inline int pthread_condattr_destroy(pthread_condattr_t *att)
...
+#define PTHREAD_MUTEX_ERRORCHECK    2
+#define PTHREAD_MUTEX_DEFAULT       PTHREAD_MUTEX_NORMAL
 
-	k_sem_init(m->sem, 1, 1);
+#define MUTEX_MAX_REC_LOCK 32767
 
-	return 0;
-}
+/* This timeout is not standard value. Set for convenience. */
+#define MEM_TIMEOUT 100
```
Comment (lpereira) [inline include/posix/pthread.h]: This is an implementation detail and shouldn't be in pthread.h.  In fact, just use K_FOREVER where the memory is allocated; the POSIX APIs make no timing guarantees when allocating a mutex, so the user knows that the thread might block forever.

File: include/posix/pthread.h [hunk 17]
```diff
@@ -157,42 +159,61 @@ static inline int pthread_condattr_destroy(pthread_condattr_t *att)
...
+ *  PTHREAD_PRIO_NONE: Ownership of mutex does not affect priority.
+ *  PTHREAD_PRIO_INHERIT: Owner's priority is boosted to the priority of
+ *      highest priority thread blocked on the mutex.
+ *  PTHREAD_PRIO_PROTECT:  Mutex has a priority ceiling.  The owner's
+ *      priority is boosted to the highest priority ceiling of all mutexes
+ *      owned (regardless of whether or not other threads are blocked on
+ *      any of these mutexes).
+ */
+#define PTHREAD_PRIO_NONE           0
+#define PTHREAD_PRIO_INHERIT        1
```
Comment (lpereira) [inline include/posix/pthread.h]: Code might test with `#ifdef PTHREAD_PRIO_INHERIT` if a certain priority level is supported by the system prior to using it.  Please use `__DEPRECATED_MACRO` or something similar to note that this macro shouldn't be used as it's not supported.

File: include/posix/pthread.h [hunk 18]
```diff
@@ -157,42 +159,61 @@ static inline int pthread_condattr_destroy(pthread_condattr_t *att)
...
-				     const pthread_mutexattr_t *att)
-{
-	ARG_UNUSED(att);
+#define PTHREAD_MUTEX_NORMAL        0
+#define PTHREAD_MUTEX_RECURSIVE     1
+#define PTHREAD_MUTEX_ERRORCHECK    2
+#define PTHREAD_MUTEX_DEFAULT       PTHREAD_MUTEX_NORMAL
 
-	k_sem_init(m->sem, 1, 1);
+#define MUTEX_MAX_REC_LOCK 32767
```
Comment (lpereira) [inline include/posix/pthread.h]: This is an implementation detail and should not be in pthread.h.

**Review by lpereira — DISMISSED** [review]

File: lib/posix/pthread_mutex.c [hunk 19]
```diff
@@ -9,16 +9,218 @@
...
+ * @brief Intialize POSIX mutex.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_init(pthread_mutex_t *m,
+				     const pthread_mutexattr_t *attr)
+{
+	const pthread_mutexattr_t *mattr;
+
+	if (k_mem_pool_alloc(&posix_mutex_sem, &m->block,
```
Comment (andrewboie) [inline lib/posix/pthread_mutex.c]: why use k_mem_pool instead of sys_mem_pool? I would vastly prefer that you use sys_mem_pool for mutex allocations. for one thing, sys_mem_pool is user mode compatible and k_mem_pool never will be.
Also please name this something else, the _sem suffix suggests that this object is a semaphore and it's not.

File: lib/posix/pthread_mutex.c (hunk 2 above)
Comment (andrewboie) [reply lib/posix/pthread_mutex.c]: please use a sys_mem_pool instead of a k_mem_pool

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]

File: tests/posix/mutex/prj.conf [hunk 20]
```diff
@@ -0,0 +1,4 @@
+CONFIG_PTHREAD_IPC=y
+#CONFIG_MAIN_THREAD_PRIORITY=2
```
[the remark this answers is not in the document]
Comment (punitvara) [reply tests/posix/mutex/prj.conf]: Done

File: include/posix/pthread.h (hunk 16 above)
Comment (punitvara) [reply include/posix/pthread.h]: Done

File: lib/posix/pthread_mutex.c [hunk 21]
```diff
@@ -9,16 +9,218 @@
...
+
+/**
+ * @brief Read protocol attribute for mutex.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutexattr_getprotocol(const pthread_mutexattr_t *attr,
+				  int *protocol)
+{
+	*protocol = attr->protocol;
```
[the remark this answers is not in the document]
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: Done

File: lib/posix/pthread_mutex.c [hunk 22]
```diff
@@ -9,16 +9,218 @@
...
+int pthread_mutexattr_setprotocol(pthread_mutexattr_t *attr, int protocol)
+{
+	/* FIXME: Return error for all non PRIO_NONE protocols
+	 * because they are not supported at the moment.
+	 */
+	if (protocol != PTHREAD_PRIO_NONE) {
+		return EINVAL;
+	}
+
+	attr->protocol = protocol;
```
[the remark this answers is not in the document]
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: Done

File: include/posix/pthread.h (hunk 17 above)
Comment (punitvara) [reply include/posix/pthread.h]: Done

File: lib/posix/pthread_mutex.c (hunk 19 above)
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: Sure. I have used SYS_MEM_POOL and passed similar arguments as k_mem_pool_alloc but its failing. Is there any calculation I need to consider for MAX , MIN blocks here ? @andrewboie

File: lib/posix/pthread_mutex.c [hunk 23]
```diff
@@ -9,16 +9,218 @@
...
+	key = irq_lock();
+	m->lock_count--;
+
+	if (m->lock_count == 0) {
+		m->owner = NULL;
+	}
 	irq_unlock(key);
 
-	return ret;
+	k_sem_give(m->sem);
```
[the remark this answers is not in the document]
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: I am tracking count separately and using semaphore to block the activity/event. IRQs are used to protect shared resource. @lpereira

File: lib/posix/pthread_mutex.c [hunk 24]
```diff
@@ -9,16 +9,218 @@
...
+	int rc = 0, type, count;
+	pthread_t owner;
+	int key = irq_lock();
+
+	owner = m->owner;
+	type = m->type;
+	count = m->lock_count;
+	irq_unlock(key);
+
+	if (owner == pthread_self()) {
```
[the remark this answers is not in the document]
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: I though each thread has local variables so even another thread can't access local variable of another thread as they always will be on different stack. However, I changed full function under one IRQ lock.

File: lib/posix/pthread_mutex.c [hunk 25]
```diff
@@ -9,16 +9,218 @@
...
+	}
+
+	if (type == PTHREAD_MUTEX_RECURSIVE) {
+
+		if ((owner == pthread_self()) || (count == 0)) {
+			if (count >= MUTEX_MAX_REC_LOCK) {
+				return EAGAIN;
+			}
+			rc = -k_sem_take(m->sem, timeout);
+			count++;
```
[the remark this answers is not in the document]
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: Count is being protected by IRQ. Semaphore is just for blocking for timeout/k_FOREVER. When we take semaphore count is decreased but according to definition of mutex lock. When you are locking mutex, count should increased to 1 and so on. So I am tracking count separately.

File: lib/posix/pthread_mutex.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (lpereira) [reply lib/posix/pthread_mutex.c]: You are correct in the sense that, once a value is in the call frame in Thread A, Thread B can't change or see it.  However, decisions are still being taken on the original values inside the mutex struct.

Imagine for instance the following flux:

-  Thread A copies owner, type, and lock_count.
-  Thread A unlocks IRQs
-  Thread A's timeslice ends; the scheduler picks Thread B
-  Thread B copies owner, type, and lock_count.
-  Thread B unlocks IRQs

Now Thread B contains the same type and lock_count inside their call frames.  The owner is going to be different, of course, but all other decisions will be made with the original values.

File: include/posix/sys/types.h [hunk 26]
```diff
@@ -35,10 +35,15 @@ typedef struct k_sem sem_t;
 
 /* Mutex */
 typedef struct pthread_mutex {
+	struct k_mem_block block;
 	struct k_sem *sem;
+	pthread_t owner;
+	u32_t lock_count;
+	int type;
```
Comment (andyross) [inline include/posix/sys/types.h]: This changes makes struct pthread_mutex **six times larger**.  Is that really worth it for the features being added?  Is there no way to configure out the stuff you don't want (e.g. a kconfig that allows you to choose whether or not to support recursive locks, which would eliminate owner and count...)

File: lib/posix/pthread_mutex.c [hunk 27]
```diff
@@ -9,16 +9,199 @@
...
+		}
+	} else {
+		rc = k_sem_take(m->sem, timeout);
+		count = 1;
+	}
+
+	if (rc == 0) {
+		key = irq_lock();
+		m->owner = pthread_self();
+		m->lock_count = count;
```
Comment (andyross) [inline lib/posix/pthread_mutex.c]: Here is an example of the race @lpereira was talking about: you're setting count here to a value computed **outside** the critical section!  There is no guarantee that someone else didn't muck with the mutex while the irq lock was released, so the value you computed above for "count" may be wrong here.

File: lib/posix/pthread_mutex.c [hunk 28]
```diff
@@ -9,16 +9,199 @@
...
+	irq_unlock(key);
+
+	if (owner != pthread_self()) {
+		return EPERM;
+	}
+
+	if (count == 0) {
+		return EINVAL;
+	}
+
```
Comment (andyross) [inline lib/posix/pthread_mutex.c]: There's a race here too.  Consider two threads calling pthread_mutex_unlock() incorrectly, on a lock that was locked only once.  Both of them get to this location where I added the comment, decide to decrement the lock, and continue on.  Now both threads have completed successfully when one of them should have returned an error.   And worse, lock_count is now -1 and future calls are probably going to explode in hilarious ways.

**Review by andyross — CHANGES_REQUESTED** [review]
The races have to be fixed for sure.

And... this is a **lot** of bloat.  Do we really want this?   For reference, the sizeof(pthread_mutex_t) after this change is the same as it is in ia32 glibc.  And we have the extra overhead involved in the k_sem which is referenced by pointer, making us bigger than linux.   Is it really OK that our pthread_mutex_t is bigger than linux?

File: include/posix/sys/types.h (hunk 26 above)
Comment (punitvara) [reply include/posix/sys/types.h]: We can remove k_block and replace sem pointer with sem object i.e. struct k_sem sem; Also u32_t lock_count to u16_t lock_count. Would that be fine ? @andyross  I will fix other comments as well.

**Comment by ramakrishnapallala** [comment]
@andyross @lpereira 
I think the reason for race condition here is that we are using irq_lock and k_sem_take togeather. From the current implementation, we are using k_sem_take for blocking the calling thread and using  irq_lock to protect the owner and lock_count variables, with this approach(irq_lock + sem) we cannot solve both race condition and deadlock scenarios.
I think it’s better to get rid of the semaphore and use irq_lock to protect the mutex data and use _pend_current_thread() to block the calling thread. With this approach, we can avoid race and deadlock conditions and the size of the mutex will be lesser than the original.

Punit will push the changes, Please have a look into the changes and let us know the feedback.

File: lib/posix/pthread_mutex.c [hunk 29]
```diff
@@ -9,16 +9,186 @@
...
+};
+
+static int acquire_mutex(pthread_mutex_t *m, int timeout)
+{
+	int rc = 0;
+	int key = irq_lock();
+
+	if (m->owner == pthread_self()) {
+		if (m->type == PTHREAD_MUTEX_ERRORCHECK) {
+			return EDEADLK;
```
Comment (andyross) [inline lib/posix/pthread_mutex.c]: Returns without releasing the irq_lock() taken above

File: lib/posix/pthread_mutex.c [hunk 30]
```diff
@@ -9,16 +9,186 @@
...
+		}
+	}
+
+	if (m->type == PTHREAD_MUTEX_RECURSIVE) {
+
+		if ((m->owner == pthread_self()) || (m->lock_count == 0)) {
+			if (m->lock_count >= MUTEX_MAX_REC_LOCK) {
+				return EAGAIN;
```
Comment (andyross) [inline lib/posix/pthread_mutex.c]: Returns without releasing irq_lock

File: lib/posix/pthread_mutex.c [hunk 31]
```diff
@@ -9,16 +9,186 @@
...
+			}
+			rc = -k_sem_take(m->sem, timeout);
+			m->lock_count++;
+		} else {
+			rc = EPERM;
+		}
+	} else {
+		rc = k_sem_take(m->sem, timeout);
```
Comment (andyross) [inline lib/posix/pthread_mutex.c]: rc was negated above but not here, which one is the mistake?

File: lib/posix/pthread_mutex.c [hunk 32]
```diff
@@ -9,16 +9,186 @@
...
+	}
+
+	if (m->type == PTHREAD_MUTEX_RECURSIVE) {
+
+		if ((m->owner == pthread_self()) || (m->lock_count == 0)) {
+			if (m->lock_count >= MUTEX_MAX_REC_LOCK) {
+				return EAGAIN;
+			}
+			rc = -k_sem_take(m->sem, timeout);
+			m->lock_count++;
```
Comment (andyross) [inline lib/posix/pthread_mutex.c]: k_sem_take can fail, in which case it's not correct to increment the lock count, right?

File: include/posix/sys/types.h (hunk 26 above)
Comment (andyross) [reply include/posix/sys/types.h]: You could pack type into the same word as lock_count too.  I'd be... OK with that I guess.  But it still has the effect of making a pthread mutex a **much** heavier abstraction than the underlying semaphore, for IMHO not much benefit in typical code.  (Let's be honest: "no one" actually uses recursive mutexes in real world code, they're an obscure feature specified to meet someone's checklist, which makes them a poor fit for Zephyr).   Ideally I'd like the overhead to be eliminatable via kconfig and off by default.

**Review by andyross — CHANGES_REQUESTED** [review]
Don't see any obvious races, but think I found a few other errors.

Ram: yes, the combination of a blocking primitive like a semaphore and an unrelated critical section gets tricky.  But in the special case of k_sem, it's actually perfectly legal to block while holding the IRQ lock (in fact the underlying pend_current_thread() API requires it).  This code does that correctly.

BUT: this usage won't port well to spinlocks, which are not (cannot be made) recursive like that.  This will have to be revistited if we want to improve the lock granularity under SMP in the future.

**Comment by punitvara** [comment]
@andyross @lpereira I have updated as per Rama's suggestions. Thank for your time @ramakrishnapallala

File: lib/posix/pthread_mutex.c (hunk 29 above)
Comment (ramakrishnapallala) [reply lib/posix/pthread_mutex.c]: please look into the updated PR. the locking is fixed.

File: lib/posix/pthread_mutex.c (hunk 30 above)
Comment (ramakrishnapallala) [reply lib/posix/pthread_mutex.c]: same here

File: lib/posix/pthread_mutex.c (hunk 31 above)
Comment (ramakrishnapallala) [reply lib/posix/pthread_mutex.c]: removed the usage of semaphore inside pthread mutex, now the flow is straightforward.

**Comment by punitvara** [comment]
I have removed usage of semaphore itself and reduced struct size of posix mutex. Hope it looks fine now. Thank a lot for review @andyross  @lpereira Let me know what do you think about new changes.

File: lib/posix/pthread_mutex.c [hunk 33]
```diff
@@ -9,16 +9,196 @@
...
+
+/**
+ * @brief Destroy POSIX mutex.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_destroy(pthread_mutex_t *m)
+{
+	ARG_UNUSED(m);
+	return 0;
```
Comment (andyross) [inline lib/posix/pthread_mutex.c]: POSIX allows this behavior (i.e. deadlocking blocked threads) by making the result undefined if the mutex is locked, but given that you have a wait_q handy a call to _unpend_all() wouldn't be a bad idea here.

**Review by andyross — APPROVED** [review]
This looks much cleaner when build on top of a wait_q.  No really significant complaints other than the one I've already made about code bloat (which isn't academic!  This change literally blows through the existing memory budget and you had to disable the posix test on small platforms to work around that!).

File: tests/posix/mutex/src/posix_mutex.c [hunk 34]
```diff
@@ -0,0 +1,158 @@
...
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <errno.h>
+#include <pthread.h>
+
+#define STACK_SIZE 1024
```
Comment (andrewboie) [inline tests/posix/mutex/src/posix_mutex.c]: should be (1024 + CONFIG_TEST_EXTRA_STACKSIZE)

File: tests/posix/mutex/src/posix_mutex.c [hunk 35]
```diff
@@ -0,0 +1,158 @@
...
+
+static K_THREAD_STACK_DEFINE(stack, STACK_SIZE);
+
+#define SLEEP_MS 100
+
+pthread_mutex_t mutex1;
+pthread_mutex_t mutex2;
+
+void *foo(void *p1)
```
Comment (andrewboie) [inline tests/posix/mutex/src/posix_mutex.c]: please do not call these functions "foo"
Describe what they are doing.

File: tests/posix/mutex/src/posix_mutex.c [hunk 36]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	zassert_false(rc, "try lock failed\n");
+	TC_PRINT("mutex lock is taken\n");
+	zassert_false(pthread_mutex_unlock(&mutex1),
+		      "mutex unlock is falied\n");
+	return NULL;
+}
+
+void *foo2(void *p1)
```
Comment (andrewboie) [inline tests/posix/mutex/src/posix_mutex.c]: See above

File: tests/posix/mutex/prj.conf [hunk 37]
```diff
@@ -0,0 +1,3 @@
+CONFIG_PTHREAD_IPC=y
+CONFIG_ZTEST=y
+CONFIG_SEM_VALUE_MAX=32767
```
Comment (andrewboie) [inline tests/posix/mutex/prj.conf]: test case does not use a semaphore so why is this here?

File: lib/posix/pthread_mutex.c [hunk 38]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017 Intel Corporation
+ * Copyright (c) 2018 Intel Corporation
```
Comment (andrewboie) [inline lib/posix/pthread_mutex.c]: AFAIK don't need to update the year on these every time you touch a file.

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]

File: tests/posix/mutex/src/posix_mutex.c [hunk 39]
```diff
@@ -0,0 +1,158 @@
...
+		      "mutex is not taken 2nd time");
+	TC_PRINT("mutex lock is taken\n");
+	zassert_false(pthread_mutex_unlock(&mutex2),
+		      "mutex is not unlocked");
+	zassert_false(pthread_mutex_unlock(&mutex2),
+		      "mutex is not unlocked");
+	return NULL;
+}
+
+static void test_mut(void)
```
Comment (nashif) [inline tests/posix/mutex/src/posix_mutex.c]: give this a proper name please

File: tests/posix/mutex/src/posix_mutex.c [hunk 40]
```diff
@@ -0,0 +1,158 @@
...
+		if (rc == 0) {
+			break;
+		}
+		k_sleep(SLEEP_MS);
+	}
+
+	zassert_false(rc, "try lock failed\n");
+	TC_PRINT("mutex lock is taken\n");
+	zassert_false(pthread_mutex_unlock(&mutex1),
+		      "mutex unlock is falied\n");
```
Comment (nashif) [inline tests/posix/mutex/src/posix_mutex.c]: in this and all asserts below and above, remove the "\n" from the message, this is now being added by default.

File: tests/posix/mutex/src/posix_mutex.c [hunk 41]
```diff
@@ -0,0 +1,158 @@
...
+	pthread_mutex_lock(&mutex1);
+
+	zassert_equal(type, PTHREAD_MUTEX_NORMAL,
+		      "mutex type is not normal\n");
+
+	zassert_equal(protocol, PTHREAD_PRIO_NONE,
+		      "mutex protocol is not prio_none\n");
+	ret = pthread_create(&thread_1, &attr, &foo, NULL);
+
+	if (ret) {
```
Comment (nashif) [inline tests/posix/mutex/src/posix_mutex.c]: shouldn't we assert here? Why would the creation of the thread fail?

File: tests/posix/mutex/src/posix_mutex.c [hunk 42]
```diff
@@ -0,0 +1,158 @@
...
+		TC_PRINT("Thread1 creation failed %d\n", ret);
+	}
+	k_sleep(SLEEP_MS);
+	pthread_mutex_unlock(&mutex1);
+
+	pthread_join(thread_1, NULL);
+	temp = pthread_mutex_destroy(&mutex1);
+	zassert_false(temp, "Destroying mutex is failed\n");
+}
+
```
Comment (nashif) [inline tests/posix/mutex/src/posix_mutex.c]: add a brief documentation using doxygen describing what you are testing. @spoorthik  can help.

File: tests/posix/mutex/src/posix_mutex.c [hunk 43]
```diff
@@ -0,0 +1,158 @@
...
+	zassert_false(pthread_mutex_lock(&mutex2),
+		      "mutex is not taken 2nd time");
+	TC_PRINT("mutex lock is taken\n");
+	zassert_false(pthread_mutex_unlock(&mutex2),
+		      "mutex is not unlocked");
+	zassert_false(pthread_mutex_unlock(&mutex2),
+		      "mutex is not unlocked");
+	return NULL;
+}
+
```
Comment (nashif) [inline tests/posix/mutex/src/posix_mutex.c]: add a brief documentation using doxygen describing what you are testing. @spoorthik  can help.

**Review by nashif — CHANGES_REQUESTED** [review]

File: tests/posix/mutex/src/posix_mutex.c (hunk 43 above)
Comment (spoorthik) [reply tests/posix/mutex/src/posix_mutex.c]: @punitvara : Please refer this #7539

File: lib/posix/pthread_mutex.c (hunk 33 above)
Comment (punitvara) [reply lib/posix/pthread_mutex.c]: Ok Andy we will take this as an enhancement and work on the overall POSIX support later.

**Comment by punitvara** [comment]
@andrewboie @nashif  Updated changes as suggested. Thank you for feedback guys
